include/exadg/solvers_and_preconditioners/nonlinear_solvers/newton_solver.h

```diff
@@ -50,45 +50,63 @@ class Solver
   {
   }
 
-  std::tuple<unsigned int /* Newton iter */, unsigned int /* accumulated linear iter */>
+  std::tuple<unsigned int /* nonlinear iter */, unsigned int /* accumulated linear iter */>
   solve(VectorType & solution, UpdateData const & update)
   {
-    unsigned int newton_iterations = 0, linear_iterations = 0;
+    unsigned int nonlinear_iterations = 0, linear_iterations = 0;
 
-    VectorType residual, increment, temporary;
-    residual.reinit(solution);
-    increment.reinit(solution);
-    temporary.reinit(solution);
+    VectorType residual_rhs, nonlinear_solver_output, temporary_solution;
+    residual_rhs.reinit(solution);
+    nonlinear_solver_output.reinit(solution);
+    temporary_solution.reinit(solution);
 
     // evaluate residual using initial guess of solution
-    nonlinear_operator.evaluate_residual(residual, solution);
+    nonlinear_operator.evaluate_residual(residual_rhs, solution);
 
-    double norm_r   = residual.l2_norm();
+    double norm_r   = residual_rhs.l2_norm();
     double norm_r_0 = norm_r;
 
     while(norm_r > this->solver_data.abs_tol && norm_r / norm_r_0 > solver_data.rel_tol &&
-          newton_iterations < solver_data.max_iter)
+          nonlinear_iterations < solver_data.max_iter)
     {
-      // reset increment
-      increment = 0.0;
+      // reset iterate and rhs of nonlinear solver
+      if(solver_data.nonlinear_solver_type == SolverType::Newton)
+      {
+        // zero increment as initial guess for Newton
+        nonlinear_solver_output = 0.0;
+
+        // overwrite residual with right-hand side for Newton solver
+        // multiply by -1.0 since the linearized problem is
+        // "linear_operator * increment = - residual"
+        residual_rhs *= -1.0;
+      }
+      else if(solver_data.nonlinear_solver_type == SolverType::Picard)
+      {
+        // last solution as initial guess for Picard solver
+        nonlinear_solver_output = solution;
```

Comment:
I am not convinced that this "if-else-logic" is how we should implement a Picard solver. Generally, one should be able to introduce a Picard solver without touching every application. The current realization does not seem to be modular in this regard.

Reply:
If am of a different opinion, because making changes will become more and more difficult. The present solution is in my opinion a quick solution but not a clean design.

Reply:
So we should/could cook up a design for nonlinear solvers in general in base/derived classes. From my perspective, I can only think of solvers similar to Newton & Picard, since I am not familiar with other methods. I assume that other methods might even need other functionality than what we have in, say,  'convective_operator.h/.cpp'. I can generalize the nonlinear solver, but still targeting Newton or Picard.